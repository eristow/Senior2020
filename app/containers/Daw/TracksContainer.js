import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import Track from './Track';
import { bufferResource } from './bufferResource';

const Container = styled.div`
  width: 100%;
  margin-top: 10px;
`;

const TracksContainer = ({
  config,
  playing,
  currentStep,
  stepState,
  sounds,
  setBuffers,
}) => (
  <Container>
    {config.map(t => (
      <Track
        key={t}
        name={config[config.indexOf(t)]}
        num={config.indexOf(t)}
        playing={playing}
        currentStep={currentStep}
        stepState={stepState}
        buffer={bufferResource.read(sounds[t])}
        setBuffers={setBuffers}
      />
    ))}
  </Container>
);

TracksContainer.propTypes = {
  config: PropTypes.array,
  playing: PropTypes.bool,
  currentStep: PropTypes.number,
  stepState: PropTypes.object,
  sounds: PropTypes.object,
  setBuffers: PropTypes.func,
};

export default TracksContainer;
