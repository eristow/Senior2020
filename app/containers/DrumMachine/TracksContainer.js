import React, { useRef, useState, useEffect } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

import { bufferResource } from './bufferResource';
import Track from './Track';

const Wrapper = styled.div`
  display: flex;
  flex: 1;
  flex-direction: column;
  border: 2px solid #444;
  border-radius: 4px;
  margin: 0px 20px 20px;
  position: relative;
`;

const IndicatorWrapper = styled.div`
  z-index: 9;
  position: absolute;
  width: 570px;
  top: 0px;
  right: 0px;
  left: 99px;
  height: 100%;
  padding: 4px 2px;
  pointer-events: none;
`;

export default function TracksContainer({
  config,
  playing,
  currentStep,
  setBuffers,
  size,
}) {
  const trackRef = useRef();
  const [stepWidth, setStepWidth] = useState(0);
  console.log(stepWidth);

  const StepIndicator = styled.div.attrs(propsStep => ({
    step: propsStep.step,
  }))`
    position: absolute;
    top: 0px;
    left: ${propsStep => propsStep.step * stepWidth}px;
    width: ${() => stepWidth - 4}px;
    height: 100%;
    background: #00ff0040;
  `;

  useEffect(() => {
    if (trackRef.current) {
      const width = (trackRef.current.clientWidth - 97) / 16;
      setStepWidth(width);
    } else {
      setStepWidth(0);
    }
  }, [size.width]);

  return (
    <Wrapper>
      <IndicatorWrapper>
        {playing && <StepIndicator step={currentStep} />}
      </IndicatorWrapper>
      <div ref={trackRef}>
        {config.tracks.map(t => (
          <Track
            name={t}
            key={t}
            buffer={bufferResource.read(config.samples[t])}
            setBuffers={setBuffers}
          />
        ))}
      </div>
    </Wrapper>
  );
}

TracksContainer.propTypes = {
  config: PropTypes.object,
  playing: PropTypes.bool,
  currentStep: PropTypes.number,
  setBuffers: PropTypes.func,
  size: PropTypes.object,
};
