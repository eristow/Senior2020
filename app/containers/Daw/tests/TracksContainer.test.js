import React from 'react';
import { render } from 'react-testing-library';

import TracksContainer from '../TracksContainer';

describe('<TracksContainer />', () => {
  const config = ['Track1', 'Track2', 'Track3', 'Track4'];
  const tracks = [
    {
      key: 'Track1',
      name: 'Track 1',
      sound: 'https://web-daw.s3.us-east-2.amazonaws.com/kick.wav',
    },
    {
      key: 'Track2',
      name: 'Track 2',
      sound: 'https://web-daw.s3.us-east-2.amazonaws.com/snare1.wav',
    },
    {
      key: 'Track3',
      name: 'Track 3',
      sound: 'https://web-daw.s3.us-east-2.amazonaws.com/hatClosed.wav',
    },
    {
      key: 'Track4',
      name: 'Track 4',
      sound: 'https://web-daw.s3.us-east-2.amazonaws.com/hatOpen.wav',
    },
  ];

  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');

    render(
      <React.Suspense fallback={<p>loading</p>}>
        <TracksContainer config={config} tracks={tracks} />
      </React.Suspense>,
    );
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <React.Suspense fallback={<p>loading</p>}>
        <TracksContainer config={config} tracks={tracks} />
      </React.Suspense>,
    );
    expect(firstChild).toMatchSnapshot();
  });
});
