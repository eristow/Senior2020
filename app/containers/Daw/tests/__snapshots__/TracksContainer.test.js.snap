// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TracksContainer /> Should render and match the snapshot 1`] = `
<p>
  loading
</p>
`;
